src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensRangeHandler.cs

```diff
@@ -23,22 +27,53 @@ namespace Microsoft.CodeAnalysis.LanguageServer.Handler.SemanticTokens
     /// </summary>
     [ExportRoslynLanguagesLspRequestHandlerProvider(typeof(SemanticTokensRangeHandler)), Shared]
     [Method(Methods.TextDocumentSemanticTokensRangeName)]
-    internal class SemanticTokensRangeHandler : AbstractStatelessRequestHandler<LSP.SemanticTokensRangeParams, LSP.SemanticTokens>
+    internal class SemanticTokensRangeHandler : AbstractStatelessRequestHandler<LSP.SemanticTokensRangeParams, LSP.SemanticTokens>, IDisposable
     {
         private readonly IGlobalOptionService _globalOptions;
-
-        // Lock to guard _projectToCompilation dictionary
-        private readonly object _lock = new();
-        private readonly Dictionary<ProjectId, Task<Compilation?>> _projectIdToCompilation = new();
+        private readonly IAsynchronousOperationListener _asyncListener;
 
         public override bool MutatesSolutionState => false;
         public override bool RequiresLSPSolution => true;
 
+        /// <summary>
+        /// Lock over the mutable state that follows.
+        /// </summary>
+        private readonly object _gate = new();
+
+        /// <summary>
+        /// Mapping from project id to the workqueue for producing the corresponding compilation for it on the OOP server.
+        /// </summary>
+        private readonly Dictionary<ProjectId, CompilationAvailableEventSource> _projectIdToEventSource = new();
+
+        /// <summary>
+        /// Mapping from project id to the project-cone-checksum for it we were at when the project for it had its
+        /// compilation produced on the oop server.
+        /// </summary>
+        private readonly Dictionary<ProjectId, Checksum> _projectIdToLastComputedChecksum = new();
+
+        // initialized when first request comes in.
+
+        /// <summary>
+        /// Initially null.  Set to true/false when first initialized.  The other following fields will be set if this
+        /// is true.
+        /// </summary>
+        private bool? _supportsRefresh;
+
+        private LspWorkspaceManager? _lspWorkspaceManager;
+
+        /// <summary>
+        /// Debouncing queue so that we don't attempt to issue a semantic tokens refresh notification too often.
+        /// </summary>
+        private AsyncBatchingWorkQueue? _semanticTokenRefreshQueue;
+
         [ImportingConstructor]
         [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
-        public SemanticTokensRangeHandler(IGlobalOptionService globalOptions)
+        public SemanticTokensRangeHandler(
+            IGlobalOptionService globalOptions,
+            IAsynchronousOperationListenerProvider asynchronousOperationListenerProvider)
         {
             _globalOptions = globalOptions;
+            _asyncListener = asynchronousOperationListenerProvider.GetListener(FeatureAttribute.Classification);
```

Comment:
question: In the prior implementation I passed in `FeatureAttribute.Workspace` here, was that a bug/what implications does passing in `FeatureAttribute.Classification` vs `FeatureAttribute.Workspace` have?

Reply:
Great question.  The general idea of hte "Async Listener" system is to allow features to let an external system know "i'm doing async work, so if you need information from me, you should wait until that work is done".  This system exists *only* for test purposes so that tests can simulate what the user is doing, and then check for the expected results as soon as possible (e.g. instead of polling and then timing out once a certain amount of time has gone past).

The 'feature attribute' is jsut a broad way of saying what sort of functionality the async work is being done for.  that way the test can just way for that general area to be done, as opposed to waiting for async work anywhere (as there may be lots of async work happening that the test doesn't care about at all).

So here, we want to use 'Classification' so that if we have a test of LSP classification, and it only wanted to run once there was no bg-classification work, it would have a way to do it.

In general, using 'Workspace' is not a great idea as it's a very 'catch all bucket' that can mean the test has to wait a long time for everything in the workspace to settle down.

Reply:
Thanks! That was a very helpful explanation